core/src/main/scala/io/odin/loggers/AsyncLogger.scala

```diff
@@ -32,20 +32,20 @@ import cats.MonadThrow
   * Use `AsyncLogger.withAsync` to instantiate it safely
   */
 private[loggers] final class AsyncLogger[F[_]: Clock](
-    queue: Queue[F, LoggerMessage],
+    queue: Queue[F, F[Unit]],
     inner: Logger[F]
 )(
     implicit F: MonadThrow[F]
 ) extends DefaultLogger[F](inner.minLevel) {
 
   def withMinimalLevel(level: Level): Logger[F] = new AsyncLogger(queue, inner.withMinimalLevel(level))
 
-  def submit(msg: LoggerMessage): F[Unit] = queue.tryOffer(msg).void
+  def submit(msg: LoggerMessage): F[Unit] = queue.tryOffer(inner.log(msg)).void
 
-  private[loggers] def drain: F[Unit] = drainAll.flatMap(msgs => inner.log(msgs.toList)).voidError
+  private[loggers] def drain: F[Unit] = drainAll.flatMap(logActions => logActions.sequence_).voidError
 
-  private def drainAll: F[Vector[LoggerMessage]] =
-    F.tailRecM(Vector.empty[LoggerMessage]) { acc =>
+  private def drainAll: F[Vector[F[Unit]]] =
```

Comment:
Memory-wise, moving elements from the queue to the vector is ineffective. 

Also, the 'drain' step must be finite if we want to track the lag:
```scala
private def drain: F[Unit] = 
  queue.size.flatMap { initSize =>
    Monad[F].tailRecM[Int, Unit](initSize) { size =>
      if (size > 0) queue.take.flatten.as(Left(size - 1))
      else Sync[F].pure(Right(()))
    }
  }
  .voidError
  
// take all elements from the queue and evaluate them
private def drainAll: F[Unit] = 
  queue.tryTakeN(None).flatMap(_.sequence_).voidError
```

And the background consumer:
```scala
def drainLoop: F[Unit] = F.delayBy(logger.drain, timeWindow).foreverM
F.background(drainLoop).onFinalize(logger.drainAll).void
```